exporter/otlpexporter/mock_receiver.go

```diff
@@ -0,0 +1,64 @@
+// Copyright The OpenTelemetry Authors
+// SPDX-License-Identifier: Apache-2.0
+
+package otlpexporter // import "go.opentelemetry.io/collector/exporter/otlpexporter"
+
+import (
+	"context"
+	"errors"
+
+	"go.opentelemetry.io/collector/component/componenttest"
+	"go.opentelemetry.io/collector/exporter/exportertest"
+	"go.opentelemetry.io/collector/receiver"
+	"go.opentelemetry.io/collector/receiver/otlpreceiver"
+	"go.opentelemetry.io/collector/receiver/receivertest"
+)
+
+type mockOtlpReceiver struct {
+	mockConsumer    exportertest.MockConsumer
+	traceReceiver   receiver.Traces
+	metricsReceiver receiver.Metrics
+	logReceiver     receiver.Logs
+}
+
+func newOTLPDataReceiver(mockConsumer exportertest.MockConsumer) *mockOtlpReceiver {
+	return &mockOtlpReceiver{
+		mockConsumer: mockConsumer,
+	}
+}
+
+func (bor *mockOtlpReceiver) Start() error {
+	factory := otlpreceiver.NewFactory()
+	cfg := factory.CreateDefaultConfig().(*otlpreceiver.Config)
+	cfg.HTTP = nil
+	var err error
+	set := receivertest.NewNopCreateSettings()
+	if bor.traceReceiver, err = factory.CreateTracesReceiver(context.Background(), set, cfg, bor.mockConsumer); err != nil {
+		return err
+	}
+	if bor.metricsReceiver, err = factory.CreateMetricsReceiver(context.Background(), set, cfg, bor.mockConsumer); err != nil {
+		return err
+	}
+	if bor.logReceiver, err = factory.CreateLogsReceiver(context.Background(), set, cfg, bor.mockConsumer); err != nil {
+		return err
+	}
+
+	if err = bor.traceReceiver.Start(context.Background(), componenttest.NewNopHost()); err != nil {
+		return err
+	}
+	if err = bor.metricsReceiver.Start(context.Background(), componenttest.NewNopHost()); err != nil {
+		return err
+	}
+	return bor.logReceiver.Start(context.Background(), componenttest.NewNopHost())
+}
+
+func (bor *mockOtlpReceiver) Stop() error {
+	bor.mockConsumer.Clear()
+	err := bor.traceReceiver.Shutdown(context.Background())
+	err = errors.Join(err, bor.metricsReceiver.Shutdown(context.Background()))
+	return errors.Join(err, bor.logReceiver.Shutdown(context.Background()))
+}
+
+func (bor *mockOtlpReceiver) RequestCounter() exportertest.RequestCounter {
```

Comment:
The MockConsumer has all the information about received/rejected data, right? So we don't need to require implementation of this method and don't need to expose RequestCounter in the `exportertest`

Reply:
To clarify - do you mean we don't need a request counter at all? This is the only part that contain information about nonpermanent and permanent errors. According to the implementation I have now we need it exported to check if the test finished successfully:

```
reqCounter := mockReceiver.RequestCounter()
// The overall number of requests sent by exporter
fmt.Printf("Number of export tries: %d\n", reqCounter.total)
// Successfully delivered items
fmt.Printf("Total items received successfully: %d\n", reqCounter.success)
// Number of errors that happened
fmt.Printf("Number of permanent errors: %d\n", reqCounter.error.permanent)
fmt.Printf("Number of non-permanent errors: %d\n", reqCounter.error.nonpermanent)

assert.EventuallyWithT(params.T, func(c *assert.CollectT) {
	checkIfTestPassed(params.T, params.NumberOfTestElements, reqCounter)
}, 2*time.Second, 100*time.Millisecond)
```

MockConsumer indeed has all the information, but it's not exposed at all in `exportertest/contract_checker.go`, that's why I use it through `MockReceiver`. Do you suggest moving this check logic somewhere else?

Reply:
> MockConsumer indeed has all the information, but it's not exposed at all in exportertest/contract_checker.go, that's why I use it through MockReceiver.

It's available in the `reqCounter` and it's not required to be exposed because it's being used in the same package - the checker, not on the API consumer side. You can convert the `MockConsumer` interface into the `mockConsumer` struct to get access to all the private field. But that maybe too hacky. I think we can expose the struct instead of the interface to make it easier: so remove `MockReceiver` interface and expose the `mockConsumer` struct -> `MockConsumer`. Sorry, looks like I'm asking to undo what I asked before, but we need to get the the point where the API is clean :)

Reply:
Ok, so I deleted MockConsumer interface completely, exposed MockConsumer struct instead, but I didn't delete `MockReceiver` interface as I thought we'd reuse it for otlp http exporter. Now I use request counter like this: `reqCounter := mockConsumer.reqCounter` and the MockReceiver interface is:

```
type MockReceiver interface {
	Start() error
	Stop() error
}
```

Reply:
Sounds great!